extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -642,12 +642,22 @@ public void updateCounters(CounterSnapshotsTree snapshotsTree)
       // Present means the warning limit was exceeded, and warnings have therefore turned into an error.
       String errorCode = warningsExceeded.get().lhs;
       Long limit = warningsExceeded.get().rhs;
-      workerError(MSQErrorReport.fromFault(
-          id(),
-          selfDruidNode.getHost(),
-          null,
-          new TooManyWarningsFault(limit.intValue(), errorCode)
-      ));
+
+      if (limit == 0L) {
+        for (MSQErrorReport errorReport1 : workerWarnings) {
+          if (errorReport1.getFault().getErrorCode().equalsIgnoreCase(errorCode)) {
+            workerError(errorReport1);
+            break;
+          }
+        }
```

Comment:
We should also throw an illegal state exception if we donot find the error code in the workerWarnings.

Reply:
Makes sense, or else the controller would fail to stop.

Reply:
I was thinking, let the worker impl directly hit worker error endpoint in case the limit is 0 wdyt ?

Reply:
Trying out the changes, that seems like a much better alternative.